department-of-veterans-affairs/next-build
# VACMS-14599 code quality tools

## Description
Closes [#14599](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/14599).

Main changes:
- Disentangles prettier and ESLint. Prettier no longer recommends using [eslint-plugin-prettier](https://prettier.io/docs/en/integrating-with-linters), and instead suggests only using `eslint-config-prettier`, which we also had already.
- always use `yarn lint`, which is `next lint`, which is next.js's wrapper around eslint. eslint config was simplified to match a bit.
- update CI workflow to cache yarn, use the `--immutable` flag instead of `--frozen-lockfile` (does the same thing without the deprecation message), and not install husky + githooks or run the example tests in CI.
- docs added

## Testing done
lots of commits in this git history to ensure that linters & prettier were writing correctly as part of lint-staged.

CI is being tested with the actions running on this PR

## Screenshots
I use PHPStorm, this is how you would configure your editor integrations with that (checkboxes for `run on save` being the big key):
<img width="983" alt="Screenshot 2023-08-03 at 2 26 41 PM" src="https://github.com/department-of-veterans-affairs/next-build/assets/11279744/d6dc1326-a915-4746-9d18-c8d8708deed9">
<img width="982" alt="Screenshot 2023-08-03 at 2 27 00 PM" src="https://github.com/department-of-veterans-affairs/next-build/assets/11279744/bce26d42-52ed-45c9-8a48-a2ccef8b0d4d">

<img width="980" alt="Screenshot 2023-08-03 at 2 26 26 PM" src="https://github.com/department-of-veterans-affairs/next-build/assets/11279744/c1c9475f-321d-4391-8f2b-32952eb39176">


## QA steps
- CI should be passing in this PR
- locally, if editor is not integrated with prettier & eslint, you should be able to stage changes and attempt to commit them and see changes happen. for instance, try including semicolons or extra spaces. prettier will format. any obviously failing js will prevent the commit with a warning.
- if editor is integrated with tooling, you should be able to see eslint warnings & notices as you type, and no longer notice prettier warnings. prettier should format automatically when you save a file.

### Definition of Done
